cli/src/upload.rs

```diff
@@ -0,0 +1,371 @@
+    if let Some(xcresult_path) = xcresult_path {
+        let xcresult = XCResult::new(xcresult_path);
+        let junits = xcresult?
+            .generate_junits()
+            .map_err(|e| anyhow::anyhow!("Failed to generate junit files from xcresult: {}", e))?;
+        for (i, junit) in junits.iter().enumerate() {
+            let mut junit_writer: Vec<u8> = Vec::new();
+            junit.serialize(&mut junit_writer)?;
+            let junit_temp_path = junit_temp_dir
+                .path()
+                .join(format!("xcresult_junit_{}.xml", i));
+            let mut junit_temp = std::fs::File::create(&junit_temp_path)?;
+            junit_temp
+                .write_all(&junit_writer)
+                .map_err(|e| anyhow::anyhow!("Failed to write junit file: {}", e))?;
+            let junit_temp_path_str = junit_temp_path.to_str();
+            if let Some(junit_temp_path_string) = junit_temp_path_str {
+                temp_paths.push(junit_temp_path_string.to_string());
+            } else {
+                return Err(anyhow::anyhow!(
+                    "Failed to convert junit temp path to string."
+                ));
+            }
+        }
+    }
+    Ok(temp_paths)
+}
+
+fn junit_require() -> &'static str {
+    if cfg!(target_os = "macos") {
+        "xcresult_path"
+    } else {
+        "junit_paths"
+    }
+}
+
+fn parse_num_tests(file_sets: &[FileSet]) -> anyhow::Result<usize> {
+    file_sets
+        .iter()
+        .try_fold(0, |mut num_tests, file_set| -> anyhow::Result<usize> {
+            let num_tests_in_file_set = file_set.files.iter().try_fold(
+                0,
+                |mut num_tests_in_file_set, bundled_file| -> anyhow::Result<usize> {
+                    let path = std::path::Path::new(&bundled_file.original_path);
+                    let file = std::fs::File::open(path)?;
+                    let file_buf_reader = BufReader::new(file);
+                    let mut junit_parser = JunitParser::new();
+                    if let Err(e) = junit_parser.parse(file_buf_reader) {
+                        log::warn!(
+                            "Encountered error while parsing file {}: {}",
+                            bundled_file.original_path_rel,
+                            e
+                        );
+
+                        return Ok(num_tests_in_file_set);
+                    }
+
+                    let num_tests_in_file = junit_parser.into_reports().iter().fold(
+                        0,
+                        |mut num_tests_in_file, report| {
+                            num_tests_in_file += report.tests;
+                            num_tests_in_file
+                        },
+                    );
+                    num_tests_in_file_set += num_tests_in_file;
+
+                    Ok(num_tests_in_file_set)
+                },
+            )?;
+
+            num_tests += num_tests_in_file_set;
+
+            Ok(num_tests)
+        })
+}
```

Comment:
This can be much more readable with some flat mapping. Also, do we want to hard error on files we cannot read or do we want to warn on them?

```suggestion

fn parse_num_tests(file_sets: &[FileSet]) -> usize {
    file_sets
        .iter()
        .flat_map(|file_set| &file_set.files)
        .filter_map(|bundled_file| {
            let path = std::path::Path::new(&bundled_file.original_path);
            let file = std::fs::File::open(path);
            if let Err(ref e) = file {
                log::warn!(
                    "Could not open file {}: {}",
                    bundled_file.original_path_rel,
                    e
                );
            }
            file.ok().map(|f| (f, bundled_file))
        })
        .filter_map(|(file, bundled_file)| {
            let file_buf_reader = BufReader::new(file);
            let mut junit_parser = JunitParser::new();
            if let Err(e) = junit_parser.parse(file_buf_reader) {
                log::warn!(
                    "Encountered error while parsing file {}: {}",
                    bundled_file.original_path_rel,
                    e
                );

                return None;
            }
            Some(junit_parser)
        })
        .flat_map(|junit_parser| junit_parser.into_reports())
        .fold(0, |num_tests, report| num_tests + report.tests)
}
```